src/diffusers/schedulers/scheduling_ddim.py

```diff
@@ -355,5 +355,25 @@ def add_noise(
         noisy_samples = sqrt_alpha_prod * original_samples + sqrt_one_minus_alpha_prod * noise
         return noisy_samples
 
+    def get_velocity(
+        self, sample: torch.FloatTensor, noise: torch.FloatTensor, timesteps: torch.IntTensor
+    ) -> torch.FloatTensor:
+        # Make sure alphas_cumprod and timestep have same device and dtype as sample
+        self.alphas_cumprod = self.alphas_cumprod.to(device=sample.device, dtype=sample.dtype)
+        timesteps = timesteps.to(sample.device)
+
+        sqrt_alpha_prod = self.alphas_cumprod[timesteps] ** 0.5
+        sqrt_alpha_prod = sqrt_alpha_prod.flatten()
+        while len(sqrt_alpha_prod.shape) < len(sample.shape):
+            sqrt_alpha_prod = sqrt_alpha_prod.unsqueeze(-1)
+
+        sqrt_one_minus_alpha_prod = (1 - self.alphas_cumprod[timesteps]) ** 0.5
+        sqrt_one_minus_alpha_prod = sqrt_one_minus_alpha_prod.flatten()
+        while len(sqrt_one_minus_alpha_prod.shape) < len(sample.shape):
+            sqrt_one_minus_alpha_prod = sqrt_one_minus_alpha_prod.unsqueeze(-1)
+
+        velocity = sqrt_alpha_prod * noise - sqrt_one_minus_alpha_prod * sample
+        return velocity
```

Comment:
This looks very similar to `add_noise` doesn't it? Would it make sense to make both implementations rely on a common function?

Reply:
Yes, could refactor it in a follow-up PR , wdyt @patrickvonplaten

Reply:
Since we use both `add_noise` and `get_velocity` in the same training step for v-prediction, I'm ok with keeping them separate. But longer-term we might benefit from factoring out or condensing the `alpha_prod` code in both functions (everything above `velocity =`).

Reply:
Agree ! I think it's clear to add it directly to `add_noise` and then maye use of `self.config.prediction_type` - could we do this in this PR maybe?

Reply:
In favour of keeping `get_velocity` because for v-prediction we need both the velocity and noised image, so if we modify `add_noise`, we'll need to return a `tuple` as output, which will complicate it bit.

`get_velocity` is clearer to understand as it makes it clear that it's different from add_noise.